docs/source/addingsolver.rst

```diff
@@ -0,0 +1,248 @@
+a piece of code that prepares the solver prior to being able to apply a step. In general, this requires defining the
+data vector ``y``, the initial guess of the solver ``x0`` (if not provided, this will be automatically set to be a zero
+vector), and various hyperparameters of the solver - e.g., those involved in the stopping criterion. For example in
+this case we only have two parameters: ``niter`` refers to the maximum allowed number of iterations, and ``tol`` to
+tolerance on the residual norm (the solver will be stopped if this is smaller than the chosen tolerance). Moreover,
+we always have the possibility to decide whether we want to operate the solver (in this case its setup part) in verbose
+or silent mode. This is driven by the ``show`` parameter. We will soon discuss how to choose what to print on screen in
+case of verbose mode (``show=True``). The setup method can be loosely seen as composed of 3 parts. First, the data
+vector and hyperparamters are stored as members of the class. Moreover the type of the ``y`` vector is checked to
+evaluate whether to use numpy or cupy for algebraic operations (this is done by ``self.ncp = get_array_module(y)``).
+Second, the starting guess is initialized using either the provided vector ``x0`` or a zero vector. Finally, a number
+of variables are initialized to be used inside the ``step`` method to keep track of the optimization process. Moreover,
+note that the ``setup`` method returns the created starting guess ``x`` (does not store it as member of the class).
+
+.. code-block:: python
+
+    def setup(self, y, x0=None, niter=None, tol=1e-4, show=False):
+
+        self.y = y
+        self.tol = tol
+        self.niter = niter
+        self.ncp = get_array_module(y)
+
+        # initialize solver
+        if x0 is None:
+            x = self.ncp.zeros(self.Op.shape[1], dtype=self.y.dtype)
+            self.r = self.y.copy()
+        else:
+            x = x0.copy()
+            self.r = self.y - self.Op.matvec(x)
+        self.c = self.r.copy()
+        self.kold = self.ncp.abs(self.r.dot(self.r.conj()))
+
+        # create variables to track the residual norm and iterations
+        self.cost = []
+        self.cost.append(np.sqrt(self.kold))
+        self.iiter = 0
+
+        # print setup
+        if show:
+            self._print_setup(np.iscomplexobj(x))
+        return x
+
+At this point, we need to implement the core of the solver, its `step`. Here, we take the input at previous iterate,
+update it following the rule of the solver of choice, and return it. The other input parameter required by this method
+is ``show`` to choose whether we want to print a report of the step on screen or not. However, if appropriate, a user
+can add additional input parameters. For CG, the step is:
+
+.. code-block:: python
+
+    def step(self, x, show=False):
+        Opc = self.Op.matvec(self.c)
+        cOpc = self.ncp.abs(self.c.dot(Opc.conj()))
```

Comment:
I deliberately use matvec/rmatvec when possible inside the solver because they avoid the various checks that dot does (which is triggered by @). And you can see this in the tutorial Linear Operators. Do you think this is a wise strategy or an overkill? 

Maybe I should explain this in the documentation instead of switching to @?

Reply:
Ok, let's keep it matvec for now. I was mostly thinking about when we need to extend it for ndarray support.